standardize signature for Index reductions, implement nanmean for datetime64 dtypes

jbrockmendel opened the pull request:
Same basic goal as #23890, but focusing on nanops (a module I'm not all that familiar with) first, with the array methods later, instead of vice-versa.

I think the only user-facing change (i.e. need for whatsnew) is that `Series[datetime64].mean()` now works.

I need to see if there are any more tests from #23890 or #24024 that can be copied here.  

- [x] closes #24265
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

jbrockmendel committed:
standardize signature for Index reductions, implement nanmean for datetime64 dtypes

pep8speaks commented:
Hello @jbrockmendel! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`pandas/core/base.py`](https://github.com/jbrockmendel/pandas/blob/f38ffe1b7f36d356f8313b50d60d819acec2a3f2/pandas/core/base.py) !

 - There are no PEP8 issues in the file [`pandas/core/dtypes/missing.py`](https://github.com/jbrockmendel/pandas/blob/f38ffe1b7f36d356f8313b50d60d819acec2a3f2/pandas/core/dtypes/missing.py) !

 - There are no PEP8 issues in the file [`pandas/core/indexes/datetimelike.py`](https://github.com/jbrockmendel/pandas/blob/f38ffe1b7f36d356f8313b50d60d819acec2a3f2/pandas/core/indexes/datetimelike.py) !

 - There are no PEP8 issues in the file [`pandas/core/indexes/range.py`](https://github.com/jbrockmendel/pandas/blob/f38ffe1b7f36d356f8313b50d60d819acec2a3f2/pandas/core/indexes/range.py) !

 - There are no PEP8 issues in the file [`pandas/core/nanops.py`](https://github.com/jbrockmendel/pandas/blob/f38ffe1b7f36d356f8313b50d60d819acec2a3f2/pandas/core/nanops.py) !

 - There are no PEP8 issues in the file [`pandas/tests/series/test_analytics.py`](https://github.com/jbrockmendel/pandas/blob/f38ffe1b7f36d356f8313b50d60d819acec2a3f2/pandas/tests/series/test_analytics.py) !

 - There are no PEP8 issues in the file [`pandas/tests/test_nanops.py`](https://github.com/jbrockmendel/pandas/blob/f38ffe1b7f36d356f8313b50d60d819acec2a3f2/pandas/tests/test_nanops.py) !

jbrockmendel commented:
possibly closes #21583

jbrockmendel commented:
test_nanops is a PITA in to troubleshoot.  code de-duplication at the expense of clarity may need re-thinking at some point.

jbrockmendel committed:
suppress warnings

jreback commented:
needs. parameterization pass

jreback commented on pandas/core/base.py:
```diff
@@ -994,16 +1014,24 @@ def min(self):
         --------
         numpy.ndarray.argmin
         """
+        nv.validate_minmax_axis(axis)
```
should think about wether we can / should move these checks into the nanops themselves

jbrockmendel replied:
previously discussed, this check is only for 1-d objects where `axis` is a dummy attribute

jreback commented on pandas/core/dtypes/missing.py:
```diff
@@ -108,7 +109,7 @@ def _isna_new(obj):
         raise NotImplementedError("isna is not defined for MultiIndex")
     elif isinstance(obj, (ABCSeries, np.ndarray, ABCIndexClass,
-                          ABCExtensionArray)):
+                          ABCExtensionArray, ABCDatetimeArray)):
```
why is this needed? is ABCDatetimeArray NOT an extension array?

jbrockmendel replied:
not until #24024

jreback commented on pandas/core/dtypes/missing.py:
```diff
@@ -196,6 +197,8 @@ def _isna_ndarraylike(obj):
         else:
             values = obj
         result = values.isna()
+    elif isinstance(obj, ABCDatetimeArray):
```
why is this not handled by the clause above?

jbrockmendel replied:
not until #24024

jreback commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -286,7 +286,7 @@ def min(self, axis=None, *args, **kwargs):
             return self._na_value
 
-    def argmin(self, axis=None, *args, **kwargs):
+    def argmin(self, axis=None, skipna=True, *args, **kwargs):
```
can these inherit the doc-string?

jbrockmendel replied:
I'll take a look.  Might also get rid of `*args` while at it

jbrockmendel replied:
Hmm the docstrings here and for the corresponding methods in base.IndexOpsMixin are kind of clunky.  This may merit a separate look, @datapythonista ?

datapythonista replied:
I don't know well what's the class hierarchy and whether makes sense to inherit. But we'll have to add `Parameters`, `Returns` and `Examples` section here if this is public.

jreback commented on pandas/core/indexes/range.py:
```diff
@@ -297,12 +297,14 @@ def _minmax(self, meth):
 
-    def min(self):
+    def min(self, axis=None, skipna=True):
         """The minimum value of the RangeIndex"""
```
same. why don't these inherit the doc-string

jreback commented on pandas/core/nanops.py:
```diff
@@ -212,6 +213,10 @@ def _get_values(values, skipna, fill_value=None, fill_value_typ=None,
             mask = isna(values)
 
     dtype = values.dtype
+    if is_datetime64tz_dtype(orig_values):
```
this is a very very odd place to do this as no other dtype specific things are here

jbrockmendel replied:
Semi-agree.  Most of `nanops` is numpy dtype/ndarray specific, so making datetime64tz work is klunky.  That's why I thought it made more sense to implement these directly in DTA in #23890.  But conditional on implementing this here in nanops, this check is necessary.

jreback replied:
my point is this is in the wrong place. nanops already handles dtypes just not here.

jreback replied:
so this should go in _view_if_needed (or maybe just remove that function and inline it here is ok

jreback replied:
did you do this?

jbrockmendel replied:
Haven't yet.  view_if_needed is called ~25 lines below this.  It isn't obvious to me that it is OK to move it up to here.  If it is, then sure, its a one-liner that can be inlined.

jreback replied:
yes move it, as I said above you have dtype inference in 2 places for no reason.

jreback commented on pandas/core/nanops.py:
```diff
@@ -261,19 +266,24 @@ def _na_ok_dtype(dtype):
 def _view_if_needed(values):
     if is_datetime_or_timedelta_dtype(values):
-        return values.view(np.int64)
+        try:
```
then let's add view rather than do this

jbrockmendel replied:
also waiting on #24024.

jreback replied:
well, then this is going to have to wait for .view unless you do it here.

TomAugspurger replied:
#24024 has one, but I don't know if it'll stick around. I'm pretty strongly +1 on keeping it though.

jreback commented on pandas/core/nanops.py:
```diff
@@ -457,7 +466,8 @@ def nanmean(values, axis=None, skipna=True, mask=None):
     dtype_sum = dtype_max
     dtype_count = np.float64
-    if is_integer_dtype(dtype) or is_timedelta64_dtype(dtype):
+    if (is_integer_dtype(dtype) or is_timedelta64_dtype(dtype) or
```
use needs_i8_conversion

jbrockmendel replied:
that includes PeriodDType, which isn't implemented yet (again, #23890 gets these all-at-once by implementing reductions directly on the arrays)

jreback commented on pandas/core/nanops.py:
```diff
@@ -466,14 +476,20 @@ def nanmean(values, axis=None, skipna=True, mask=None):
         the_mean = the_sum / count if count > 0 else np.nan
 
-    return _wrap_results(the_mean, dtype)
+    fill_value = None
```
this is not appropriate to do here, rather in _wrap_results

jbrockmendel replied:
OK I guess.  That changes the behavior of _wrap_results for the 8 _other_ places that call it though

jreback replied:
see my above comments. this is completely changing the code flow.

jbrockmendel committed:
requested edits

jbrockmendel committed:
implement view

jreback commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -483,6 +483,26 @@ def _maybe_mask_results(self, result, fill_value=iNaT, convert=None):
+    # These are not part of the EA API, but we implement them because
+    # pandas currently assumes they're there.
+
+    def view(self, dtype=None):
```
are we adding to EA interface? or just here (ok either way). @TomAugspurger ?

TomAugspurger replied:
I don't think ExtensionArray.view makes sense. The result of that would (I think) depend on the memory layout of the concrete array backing the ExtesnionArray, and I think that ExtensionArray should remain agnostic to that.

jbrockmendel commented:
If everyone is happy with this, I can get started on a follow-up that will trim reductions off of #24024.

TomAugspurger commented on pandas/core/base.py:
```diff
@@ -951,22 +957,36 @@ def max(self):
         ('b', 2)
         """
+        nv.validate_minmax_axis(axis)
         return nanops.nanmax(self.values)
```
Why isn't skipna passed? Is that a followup item?

jbrockmendel replied:
I had considered this a follow-up item, but am now leaning towards fixing these now.  Good catch.

jbrockmendel replied:
This turns out to be more involved than expected, requires some changes in Series._reduce and IndexOpsMixin._reduce.  Neither of which is all that invasive, but changing them entails changing/fixing/testing behavior for other reductions I hadn't planned on doing in this PR.

Currently going through to get that all done here.  I'm also OK with doing this in two phases so as to take things out of #24024 sooner rather than later.

TomAugspurger commented on pandas/core/base.py:
```diff
@@ -951,22 +957,36 @@ def max(self):
         numpy.ndarray.argmax
         """
+        nv.validate_minmax_axis(axis)
         return nanops.nanargmax(self.values)
```
Doesn't this hit https://github.com/pandas-dev/pandas/issues/24265?

jbrockmendel replied:
Not this line, but the relevant line in indexes.datetimelike does.  With the commit just pushed, this fixed 24265

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into reduction3

jbrockmendel committed:
pass skipna, tests

jbrockmendel commented:
Updated to pass skipna in the appropriate places in Index and DatetimeIndexOpsMixin methods.  Updated a handful of tests, but I'm not confident we got thorough coverage

jbrockmendel commented:
Travis fail is hypothesis

jbrockmendel committed:
fixup isort

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into reduction3

jbrockmendel commented:
rebased

jreback commented on pandas/core/base.py:
```diff
@@ -957,10 +957,16 @@ def _ndarray_values(self):
         return not self.size
 
-    def max(self):
+    def max(self, axis=None, skipna=True):
```
use **kwargs I think is consistent with how we do this elsewhere, IOW we don't explicity list the numpy support args (basically remove axis)

jbrockmendel replied:
Adding kwargs kinda makes sense (as long as we dont ignore them silently), but taking axis out seems weird; a big part of the goal is to make the signatures match across the different classes

jbrockmendel replied:
right, the series reduction method have `axis` in their sigatures

```
>>> inspect.getargspec(pd.Series.max)
ArgSpec(args=['self', 'axis', 'skipna', 'level', 'numeric_only'], varargs=None, keywords='kwargs', defaults=(None, None, None, None))
```

jreback commented on pandas/core/base.py:
```diff
@@ -957,10 +957,16 @@ def _ndarray_values(self):
         """
         Return the maximum value of the Index.
 
+        Parameters
```
also we have similar things in pandas/core/base, these should change as well.

jreback commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -307,23 +313,28 @@ def max(self, axis=None, *args, **kwargs):
                 if i8[-1] != iNaT:
                     return self._box_func(i8[-1])
 
+            if not len(self):
```
elif

jreback commented on pandas/core/indexes/datetimelike.py:
```diff
@@ -307,23 +313,28 @@ def max(self, axis=None, *args, **kwargs):
+            if not len(self):
+                return self._na_value
+
             if self.hasnans:
```
same

jreback commented on pandas/tests/reductions/test_reductions.py:
```diff
@@ -82,19 +113,27 @@ def test_nanops(self):
         assert obj.argmin() == 1
         assert obj.argmax() == 2
+        assert obj.argmin(skipna=False) == -1
+        assert obj.argmax(skipna=False) == -1
```
really? this doesn't seem right

jbrockmendel replied:
What would you expect it to be?  For Series idxmin/idxmax would give np.nan

jreback commented on pandas/tests/reductions/test_reductions.py:
```diff
@@ -558,6 +599,8 @@ def test_minmax_nat_series(self, nat_ser):
         # GH#23282
         assert nat_ser.min() is pd.NaT
         assert nat_ser.max() is pd.NaT
```
can parameterize over skipna in a lot of the code you added to tests

jbrockmendel replied:
We just implemented test_reductions, still need to move over the reduction tests from tests.indexes and tests.frames.  After they're all in one place is when to really focus on parametrization.

jbrockmendel committed:
fixup rebase screwups

jbrockmendel committed:
move len-self checks

jbrockmendel committed:
fixup more rebase screwups

jreback commented on pandas/core/series.py:
```diff
@@ -3493,6 +3494,9 @@ def _reduce(self, op, name, axis=0, skipna=True, numeric_only=None,
         # dispatch to ExtensionArray interface
         if isinstance(delegate, ExtensionArray):
             return delegate._reduce(name, skipna=skipna, **kwds)
+        elif is_datetime64_dtype(delegate):
```
should this be first?

jbrockmendel replied:
it isn't clear to me that it would make a difference

jreback commented on pandas/tests/reductions/test_reductions.py:
```diff
@@ -56,23 +56,54 @@ def test_nanops(self):
+                    assert np.isnan(result)
+                else:
+                    assert result == -1
+
```
put this test on the list for splitting / parametrizing later

jbrockmendel replied:
agreed

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into reduction3

jbrockmendel committed:
do values viewing in one place

jbrockmendel committed:
unpack Series

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into reduction3

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into reduction3

jbrockmendel committed:
implement _extract_datetimelike_values_and_dtype

jbrockmendel committed:
fix mask

TomAugspurger commented on pandas/core/base.py:
```diff
@@ -957,10 +957,16 @@ def _ndarray_values(self):
+        Parameters
+        ----------
+        axis : {None}
+            Dummy argument for consistency with Series
```
I think @jbrockmendel is listing the set of possible values, which we do use. In this case, that set is just the single value `None`, so it looks strange.

I think

```
axis : int, optional
    For compatibility with NumPy. Only 0 or None are allowed.
```

is clearer.

jreback replied:
@TomAugspurger verbiage would be good

jreback commented:
code looks good. if you can update the doc-strings then would be good.

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into reduction3

jbrockmendel committed:
requested docstring edits, remoe duplicated view

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into reduction3

jbrockmendel commented:
I think docstring comments have been addressed; bears double-checking
